src/anyp/Uri.cc

```diff
@@ -722,27 +723,25 @@ AnyP::Uri::absolute() const
  *        After copying it on in the first place! Would be less code to merge the two with a flag parameter.
  *        and never copy the query-string part in the first place
  */
-char *
+SBuf
 urlCanonicalCleanWithoutRequest(const SBuf &url, const HttpRequestMethod &method, const AnyP::UriScheme &scheme)
 {
-    LOCAL_ARRAY(char, buf, MAX_URL);
-
-    snprintf(buf, sizeof(buf), SQUIDSBUFPH, SQUIDSBUFPRINT(url));
-    buf[sizeof(buf)-1] = '\0';
-
     // URN, CONNECT method, and non-stripped URIs can go straight out
+    SBuf out = url;
     if (Config.onoff.strip_query_terms && !(method == Http::METHOD_CONNECT || scheme == AnyP::PROTO_URN)) {
-        // strip anything AFTER a question-mark
-        // leaving the '?' in place
-        if (auto t = strchr(buf, '?')) {
-            *(++t) = '\0';
-        }
+        // strip anything after a path segment
+        // leaving the delimiter in place as a visual cue of truncation (if any)
+        static const CharacterSet pathDelims("path-delim", "?#");
+        const auto pathEnd = url.findFirstOf(pathDelims);
+        if (pathEnd != SBuf::npos)
+            out = url.substr(0, pathEnd+1);
     }
 
-    if (stringHasCntl(buf))
-        xstrncpy(buf, rfc1738_escape_unescaped(buf), MAX_URL);
-
-    return buf;
+    // check for unescaped control characters
+    // RFC 1738, 5234 and 7230 require the UTF-8 range (0x80-0xFF)
```

Comment:
In new comments (at least), please update references to RFCs (short quotes or, where not practical, specific section references are best) and remove references to UTF (which is a very complex concept not directly relevant to this code):

Obsolete RFC [1738](https://www.rfc-editor.org/rfc/rfc1738) does not mention UTF AFAICT. It uses US-ASCII. RFC [3986](https://www.rfc-editor.org/rfc/rfc3986) that obsoletes RFC 1738 mentions UTC/UTF but all the _relevant_ requirements are still based on US-ASCII. This code deals with individual bytes, not UTF multibyte sequences.

RFC [5234](https://www.rfc-editor.org/rfc/rfc5234) (Augmented BNF for Syntax Specifications) does not seem to be relevant to this comment and code. It does not mention UTF.

Obsolete RFC [7230](https://www.rfc-editor.org/rfc/rfc7230) does not mention UTF. The RFCs that replace it do not mention UTF in any relevant contexts.